Expose min fee from node

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->

adp-1143

`evaluateTransactionFee` from cardano-api uses cardano-api's ProtocolParamters. Hence,
1. cardano-api's ProtocolParameters were exposed via new network method DONE
2. evaluateMinimumFee was added to transaction interface DONE
3. impl in shelley using cardano-api function was used using cardano-api's ProtocolParameters and SealedTx DONE
4. function determining the number of witnesses was implemented  DONE 
5. try to address calculation of number of certificate wits (using ledger things) DONE

Limitations:
- not for ByronEra and not for Byron witnesses

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
enable getting node's ProtocolParameters

paweljakubas committed:
add evaluateMinimumFee to transaction interface

paweljakubas committed:
add fee evaluation without wit count

paweljakubas committed:
add impl of number of wits except certification wits

paweljakubas committed:
deal with cert when computing number of witnesses

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -766,7 +769,19 @@ mkTipSyncClient tr np onPParamsUpdate onInterpreterUpdate onEraUpdate = do
-            return (pp, sp)
+
+            ppNode <- onAnyEra
+                (error "not sure at this moment how to handle that")
```
Can we not make this `Nothing`?

paweljakubas replied:
we can. good idea

paweljakubas committed:
return Maybe Node.ProtocolParameters

paweljakubas committed:
fix ledger types and add scaffolding for fromLedgerPParams

paweljakubas committed:
add fromLedgerShelleyPParams impl

paweljakubas committed:
add fromLedgerAlonzoPParams impl

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -622,6 +626,62 @@ dummySkeleton inputCount outputs = SelectionSkeleton
+    fromCardanoLovelace minFee
+  where
+    minFee = case getSealedTxBody tx of
+        InAnyCardanoEra ShelleyEra txbody ->
```
I'm looking into this, but I don't think we need to expose the ledger txbodies here... The `Cardano.Tx body wits` pattern should be enough...

Anviking replied:
Or is it intentional for some reason?

Anviking replied:
The required `IsShelleyBasedEra era`, I think we should be able to conjure up somehow...

paweljakubas replied:
unfortunately,in the case of tx certs I cannot extract from cardano-node due to how it sneaks BuildTx there. And it turn out that doing this through ledger tx is a way...

paweljakubas replied:
feel free to try without ledger bypass if you have a while

Anviking replied:
For the record, what I had missed until our quick call was:

- We need to predict how many witnesses there will be, as the transaction is not signed.
- We cannot use [estimateTransactionKeyWitnessCount](https://github.com/input-output-hk/cardano-node/blob/0ed04ab8d0a205691219f04623a3d880545a3f34/cardano-api/src/Cardano/Api/Fees.hs#L284-L285)

Anviking committed:
Add comment about copied cardano-api functions

Anviking commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -143,6 +146,15 @@ data TransactionLayer k tx = TransactionLayer
+            -- Current protocol parameters
+        -> tx
+            -- The sealed transaction
+        -> Coin
```
Perhaps we should return `Maybe` instead of erroring in case of byron, but I think leaving it as-is is fine while still trying to fit the pieces together.

Anviking replied:
I ended up adding a `Maybe` here.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -716,6 +722,166 @@ txParametersFromPParams maxBundleSize pp = W.TxParameters
 
+--------------------------------------------------------------------------------
+-- Copied from cardano-api
+-- To be removed when once again exposed.
```
I added this note for future readers

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -622,6 +626,62 @@ dummySkeleton inputCount outputs = SelectionSkeleton
+            in Cardano.evaluateTransactionFee @Cardano.MaryEra pp txbody (witsNum txbody certNum) 0
+        InAnyCardanoEra AlonzoEra txbody ->
+            let (Cardano.ShelleyTxBody _ ledgertxbody _ _ _ _) = txbody
+                certNum = length $ Alonzo.txcerts ledgertxbody
```
I think these `certNum` could be a slight overestimate, as not all certs require witnesses (KeyReg being the exception). This should affect the result of `_evaluateMinimumFee` when Daedalus users delegate for the first time.

Another, more theoretical case would also be where all the certs require just a single key witness.

Not a big deal, but could add a comment about it.

Anviking replied:
6faea90

Anviking committed:
Add comments and Maybe

1. Make evaluateMinimumFee return `Maybe Coin` rather than being a partial
function.

2. Add some comments clarifying aspects.

Anviking reviewed (approved):
LGTM, will be useful to have `currentNodeProtocolParameters`
